### ISSUE TITLE
1.1.0-RC3 implementation of git plugin leads to large memory consumption

### ISSUE THREAD

**Issue by ascheman** [kontext-e/jqassistant-plugins#21]
For very large Git instances the current implementation of the Git scanner plugin consumes a large amount of memory and may lead to a JVM abort, cf. https://groups.google.com/d/msg/jqassistant/U_gD3_FKFD8/gwqOFdaqDwAJ. A work around is to increase the JVM Heap an PermGen space. 

A better solution could be to refactor the implementation: 
- It is probably a problem of building up a complete set of Git related objects (in particular, commits) in memory. Then they are transformed to Descriptor objects and stored in neo4j.
- Improvement idea by streaming data flow: Create Descriptors directly during traversal of the Git graph and store them immediately.

**Comment by kontext-e** [comment]
To me it seems very large Git instances are currently seldom enough for jQA Git plugin users that we can use the work around for the next weeks. My prio is to get v 1.1.0 out of the door eventually. Then we can make it better.

**Comment by ascheman** [comment]
Yes, sure, that's what I think too. You should not postpone releasing until one of us finds the time to implement the improvement. For many cases it works well and for the very large git repositories (or for reading in multiple repos) there is a work around for the memory problem. So it is now show stopper and I would appreciate to have a sub optimal release now rather than a "perfect" one in the future, in particular for the current project which likes to see the released version on Maven central.


On 11.04.2016, at 14:19, Kontext E <notifications@github.com> wrote:

> To me it seems very large Git instances are currently seldom enough for jQA Git plugin users that we can use the work around for the next weeks. My prio is to get v 1.1.0 out of the door eventually. Then we can make it better.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> 

--
Gerd Aschemann --- Veröffentlichen heißt Verändern (Carmen Thomas)
+49/173/3264070 -- gerd@aschemann.net -- http://www.aschemann.net

**Comment by ascheman** [comment]
I have re-tested the problem by scanning a number of git repositories picked from the https://github.com/spring-projects universe by random:

```
 27M    spring-batch/.git
 39M    spring-boot/.git
4.4M    spring-data-cassandra/.git
5.6M    spring-data-commons/.git
106M    spring-framework/.git
496M    spring-ide/.git
 42M    spring-integration/.git
 24M    spring-ldap/.git
8.3M    spring-petclinic/.git
1.5M    spring-reactive/.git
348K    spring-security-migrate-3-to-4/.git
3.6M    spring-security-saml/.git
 27M    spring-security/.git
 18M    spring-session/.git
7.9M    spring-statemachine/.git
 21M    spring-ws/.git
```
This makes in sum ~830 MB of repository data. I was not running into the memory problem we observed with a larger scan (not only git but also Maven) somewhere else. 

I suggest to set up a general automated integration test with a large data set to investigate the problem further and to make sure for future changes that it continuously gets tested.

**Comment by kontext-e** [comment]
Tried neo4j repository, took with -Xmx10G ~1h, DB is 3.5GB.
Scan with -Xmx20G took 20min.
Did it with 4yo Laptop Core i7, 32GB RAM, SSD
Not sure if it is too much for such amount of data.

**Comment by DirkMahler** [comment]
32GB still is quite much for a notebook, mine is still running with 8GB (ok, should better have 16GB). The question is if there's a chance to be more independent of the amount of RAM available even if Git histories become large.

**Comment by kontext-e** [comment]
Indeed, and it takes a while. So definitely nothing for every commit in CI build. So yes, it is still an open point (hence this issue is still open) and no, I don't think I'm able to do something about that in the next weeks.
